src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Cohost/CohostProjectSnapshot.cs

```diff
@@ -0,0 +1,160 @@
+        {
+            var resolver = new CompilationTagHelperResolver(_telemetryReporter);
+            return resolver.GetTagHelpersAsync(_project, GetProjectEngine(), CancellationToken.None).AsTask();
+        }, joinableTaskContext.Factory);
+
+        _importsToRelatedDocumentsLazy = new Lazy<ImmutableDictionary<string, ImmutableArray<string>>>(() =>
+        {
+            var importsToRelatedDocuments = ImmutableDictionary.Create<string, ImmutableArray<string>>(FilePathNormalizer.Comparer);
+            foreach (var document in DocumentFilePaths)
+            {
+                var importTargetPaths = ProjectState.GetImportDocumentTargetPaths(document, FileKinds.GetFileKindFromFilePath(document), GetProjectEngine());
+                importsToRelatedDocuments = ProjectState.AddToImportsToRelatedDocuments(importsToRelatedDocuments, document, importTargetPaths);
+            }
+
+            return importsToRelatedDocuments;
+        });
+    }
+
+    public ProjectKey Key => _projectKey;
+
+    public RazorConfiguration Configuration => _lazyConfiguration.Value;
+
+    public IEnumerable<string> DocumentFilePaths
+        => _project.AdditionalDocuments
+            .Where(d => d.FilePath!.EndsWith(".razor", StringComparison.OrdinalIgnoreCase) || d.FilePath.EndsWith(".cshtml", StringComparison.OrdinalIgnoreCase))
+            .Select(d => d.FilePath.AssumeNotNull());
+
+    public string FilePath => _project.FilePath!;
+
+    public string IntermediateOutputPath => FilePathNormalizer.GetNormalizedDirectoryName(_project.CompilationOutputInfo.AssemblyPath);
+
+    public string? RootNamespace => _project.DefaultNamespace ?? "ASP";
+
+    public string DisplayName => _project.Name;
+
+    public VersionStamp Version => _project.Version;
+
+    public LanguageVersion CSharpLanguageVersion => ((CSharpParseOptions)_project.ParseOptions!).LanguageVersion;
+
+    public ImmutableArray<TagHelperDescriptor> TagHelpers => _tagHelpersLazy.GetValue();
+
+    public ProjectWorkspaceState ProjectWorkspaceState => ProjectWorkspaceState.Create(TagHelpers, CSharpLanguageVersion);
+
+    public IDocumentSnapshot? GetDocument(string filePath)
+    {
+        var textDocument = _project.AdditionalDocuments.FirstOrDefault(d => d.FilePath == filePath);
+        if (textDocument is null)
+        {
+            return null;
+        }
+
+        return _documentSnapshotFactory.GetOrCreate(textDocument);
+    }
+
+    public RazorProjectEngine GetProjectEngine() => _lazyProjectEngine.Value;
+
+    public ImmutableArray<IDocumentSnapshot> GetRelatedDocuments(IDocumentSnapshot document)
+    {
+        var targetPath = document.TargetPath.AssumeNotNull();
+
+        if (!_importsToRelatedDocumentsLazy.Value.TryGetValue(targetPath, out var relatedDocuments))
+        {
+            return ImmutableArray<IDocumentSnapshot>.Empty;
+        }
+
+        using var _ = ArrayBuilderPool<IDocumentSnapshot>.GetPooledObject(out var builder);
```

Comment:
Consider using a `PooledArrayBuilder` if the result will regularly be larger than 4 items.

Reply:
Do you mean if it will be less than 4 items? If it's more than 4, doesn't the `PooledArrayBuilder` just switch to using the `ArrayBuilderPool` anyway?

Reply:
Having said that, I just copied this code as is, and it can be improved slightly by specifying a capacity at least

Reply:
Yes, sorry - I meant "less than 4 items". `PooledArrayBuilder` can help avoid starvation of the `ArrayBuilderPool` and is useful when smaller arrays are expected.

Reply:
Yep, that's my reading too, and I agree. Thanks for confirming